### PR TITLE
react 16 support

### LINKED ISSUE #913: React 16 Support
## Do you want to request a feature or report a bug?
feature
## What is the current behavior?
failing in React 16
## What is the expected behavior?
it should run seamlessly just as in React 15
## Which versions of Recharts, and which browser / OS are affected by this issue? Did this work in previous versions of Recharts?
Recharts alpha 5, React 16, chrome & safari

### PR DESCRIPTION
**Pull request by kamthamc**
Updated packages required for react 16 support, resolved few issues observed in the way.

resolved #913

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -75,9 +75,9 @@
     "babel-preset-es2015": "^6.18.0",
     "babel-preset-react": "^6.16.0",
     "chai": "^3.5.0",
-    "chai-enzyme": "^0.5.2",
+    "chai-enzyme": "^0.8.0",
     "cross-env": "^3.1.4",
-    "enzyme": "^2.6.0",
+    "enzyme": "^3.0.0",
```
Comment (ahmedelgabri) [inline package.json]: You are missing `enzyme-adapter-react-16` too https://github.com/airbnb/enzyme#upgrading-from-enzyme-2x-or-react--16 :)

Comment (kamthamc) [reply package.json]: Waiting for the chai-enzyme update.

**Comment by kamthamc** [comment]
LineCharts is failing for some reason. Any help is appreciated. Otherwise i would look at it on this weekend.

**Comment by NathanCH** [comment]
It's not required to upgrade to React Router v4 btw. React Router 3.2.x upgraded to prop-types package. But looks like you did it without issue 👍 

I can try to help with the LineChart test. Did a lot of Enzyme changes this week at work lol.

**Comment by kamthamc** [comment]
@NathanCH yes itsn't not required but to test locally i have changed. Anyway newer is better right. Thanks you for helping with Linechart

**Comment by dmr** [comment]
I really like this pull request :) Thank you.

I read through your code @kamthamc and noticed you put "react-transition-group" into peerDependencies but the changelog for 1.0.0-alpha.5 states:
```
- remove react-transition-group in peer dependencies
```
--> Maybe remove this again from peerDependencies?

**Comment by dmr** [comment]
I feel like there are lots of things happening in this pull request.
I.e. the eslint updates could happen in a different step. They introduce lots of not so nice .eslintrc details IMO.

**Comment by kamthamc** [comment]
@dmr i am not touching eslint fixes, i am just upgrading libs

**Comment by kamthamc** [comment]
@dmr i can revert the commits 
60b3730, 38b788a if you think too much is going on this pr

**Comment by dmr** [comment]
I added some code here: https://github.com/kamthamc/recharts/pull/1 to fix brush behavior.

**Comment by dmr** [comment]
I don't think there is going on too much yet, but just wanted to ask how others think about this.
Thank you for pushing the react-16 issue, I really want to see this get merged once it runs.

**Comment by ahmedelgabri** [comment]
To fix the test you need to do this where you there `.find(DOMselector)` 

```diff
- expect(wrapper.find('.recharts-tooltip-cursor').length).to.equal(0);
+ expect(wrapper.find('.recharts-tooltip-cursor').hostNodes().length).to.equal(0);
```

The behaviour of `.find()` changed in Enzyme 3.x, more info [here](https://github.com/airbnb/enzyme/issues/1237), [here](https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md#element-referential-identity-is-no-longer-preserved) & [here](https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/hostNodes.md)

As for the changes in general, I think this PR should focus only on React 16 upgrade. ESLint, React Router, etc.. (if they are not needed for R16) should be done in separate PRs. To keep the change surface to the minimum & we keep the PR focused.

One thing that is worrying that I'm not seeing any activity from the owners of the library & not even a comment.

**Comment by xile611** [comment]
@ahmedelgabri Thanks to all contributors, I have review part of the codes.  It seems no big problem, I'll review line-by-line once the test cases are passed.

**Comment by dmr** [comment]
There ist still some testcases failing but maybe it's a good starting point

**Comment by ahmedelgabri** [comment]
For the failing tests, most probably the answers are here https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md because enzyme went through a big change & some APIs changed in behaviour

**Comment by kamthamc** [comment]
@ahmedelgabri i tried with hostnodes and its undefined. I was going through documentation & i noticed i have to increase the length by 1.

**Comment by ahmedelgabri** [comment]
That's odd I just cloned the repo & tried `hostNodes()` & it worked fine, I don't think increasing by one will work because it's not

**Comment by ahmedelgabri** [comment]
@kamthamc I opened a PR to merge into your branch that fixes the failing test

**Comment by ryuken** [comment]
any eta when this will be merged? =p

**Comment by ahmedelgabri** [comment]
I fixed a couple of more places that required `.hostNodes()` the tests are still failing, maybe @xile611 can give us some hints or help on how to fix them.

**Comment by ahmedelgabri** [comment]
Here are a couple of places were the tests are failing:

https://github.com/kamthamc/recharts/blob/739ca2ea03eab84c931368b733f6c2c77feb1f88/test/specs/chart/LineChartSpec.js#L340

This now returns `6` instead of `3` & you can easily see this if you do `console.log(newLineDots.children().debug())`, on master it returns `3` though.

https://github.com/kamthamc/recharts/blob/739ca2ea03eab84c931368b733f6c2c77feb1f88/test/specs/chart/LineChartSpec.js#L343

These are not the same objects

```js
console.log(newLineDots.childAt(0).props().payload)
console.log(data[2])
```

Will give this output
```
LOG: '{
  "name": "Page A",
  "uv": 400,
  "pv": 2400,
  "amt": 2400
}'
LOG: '{
  "name": "Page C",
  "uv": 300,
  "pv": 1398,
  "amt": 2400
}'
```

**Comment by dmr** [comment]
What is missing here to merge this pull request?

**Comment by ahmedelgabri** [comment]
@xile611 any updates on this, tests are passing & seems like everything is good, no?

**Comment by xile611** [comment]
@kamthamc  @ahmedelgabri @dmr  Thanks! Since it's a break change that recharts don't support React 15, I'll release a new version 1.0.0-beta.0 after merging this PR. Do you have any other idea?

**Comment by ahmedelgabri** [comment]
@xile611 makes total sense, thanks!

**Comment by dmr** [comment]
I agree. 
Maybe offtopic, but what else is blocking a 1.0.0?

**Comment by declanelcocks** [comment]
Also been testing this PR against an existing project with extensive `recharts` usage. No issues found 👍

**Comment by ahmedelgabri** [comment]
Any update @xile611, this is really the only thing blocking us to update to React 16 in our apps.

**Pull request merged**